Limit the subquery results by memory usage

LakshSingla opened the pull request:
### Description

### Overview

Currently, in the ClientQuerySegmentWalker, when the data sources get inlined, they can be limited by the number of rows to prevent a query (subquery) from hogging up the broker's memory. This however doesn't have a proper correspondence with the memory used, since a row can have multiple columns with varying amounts of data in them. Therefore it would be better if a memory limit is also available, which prevents the subquery's results from exploding beyond a certain memory limit.

This PR aims to use the Frame which was introduced along with the MSQ to store the inline results. Since the Frames are backed by memory, we can fetch the memory used by the frame, and correspondingly the data source to estimate the size that is consumed by the inline data source. This is a close estimate of the size taken by the subquery results.

### Configuration
1. User can set the `maxSubqueryBytes` key in the query context with the value that is the upper bound on the number of bytes the subquery's results can take.
2. Often, the result types of the query's results are unknown to the broker. Materializing to frames require that the types be present, which can cause results to not be materialized. Therefore there is an additional undocumented parameter `useNestedForUnknownTypeInSubquery` parameter that by default is set to false but can be set to true, which controls how to handle the column's whose types are unknown (i.e. the columnType is empty/null). If the parameter is set, then the null types are serded as nested JSON data, which should handle most of the common cases.

### Behaviour
As proposed in the [review comment](https://github.com/apache/druid/pull/13952#pullrequestreview-1409574208) if
* if `maxSubqueryBytes` is not set
    * We use `maxSubqueryRows` to limit the results of the subquery by the number of rows. Also, we execute the older code path, which doesn't materialize the results to frames
* if `maxSubqueryBytes` is set
  * if we can materialize the results of all the subqueries to frames
    * We limit the results of the subquery by the number of bytes
   * if we can't materialize the results of all the subqueries to frames
     * We default to the old code path and limit the results of the subquery by the number of rows (if set). 

### Proposed changes
1. ClientQuerySegmentWalker#toInlineDataSource now has distinct code paths which convert the results respectively to frames (new changes) or the iterable of rows (old changes)
2. The query tool chests have been updated to materialize the results as frames if possible.

### Supporting changes
1. `FrameBasedIndexedTable` - An indexed table that works on the `FrameBasedInlineDataSource`. It indexes the key columns and provides a way to extract the columnReader for the columns of the data source
3. `FrameBasedInlineDataSource` - Inline data source which is based upon an underlying list of frames. Frames can be written using individual row signatures. The datasource itself has its own signature which is "cumulative" signature of the underlying frames.
4. `IterableRowsCursorHelper` - Creates a cursor from an iterable representing the rows of a datasource.
5. `ConcatCursor` - Cursor representing concatenation of multiple underlying cursors.
6. ScanResultValue has been updated to provide the row signature whenever present in the underlying datasources.
7. `FrameBasedInlineDataSourceSerializer` - Serializes a `FrameBasedInlineDataSource` as if it was the traditional InlineDataSource. This is done to keep the communication protocol between the broker and the data servers (historicals) unchanged.

### Testing
* The set of changes has been tested on the existing CalciteQueryTests stack.
* The changes have been tested on a local deployment

### Impact on existing deployments
1. This change shouldn't affect the existing deployments. The communication protocol between the broker and the historical is unchanged (the frame-based datasources, inline as existing traditional datasources).
2. The change to `ScanResultValue` shouldn't affect the upgrade process since historicals are updated before the brokers in the proper upgrade method, therefore brokers can always assume that the `ScanResultValue` should contain the updated information (of signature). Broker's currently don't break even if the fetched `ScanResultvalue` has rowSignature as null, therefore existing deployments shouldn't be affected
3. The new code paths are only deployed when the memory limit is mentioned by the user in the queries, therefore existing queries should continue to work as is, without any performance impacts. 

### Follow up
1. Create an automatic user-friendly configuration based on the [guidelines](https://github.com/apache/druid/pull/13952#pullrequestreview-1409574208).

#### Release note
Users can now add a guardrail to prevent subquery’s results from exceeding the set number of bytes by setting `druid.server.http.maxSubqueryRows` in Broker's config or `maxSubqueryRows` in the query context. This feature is experimental for now and would default back to row-based limiting in case it fails to get the accurate size of the results consumed by the query.
 
<hr>

<!-- Check the items by putting "x" in the brackets for the done things. Not all of these items apply to every PR. Remove the items which are not done or not relevant to the PR. None of the items from the checklist below are strictly necessary, but it would be very helpful if you at least self-review the PR. -->

This PR has:

- [x] been self-reviewed.
   - [x] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [x] added documentation for new or modified features or behaviors.
- [x] a release note entry in the PR description.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

cryptoe commented on processing/src/main/java/org/apache/druid/query/InlineDataSource.java:
```diff
@@ -53,14 +63,17 @@
 public class InlineDataSource implements DataSource
 {
   private final Iterable<Object[]> rows;
+  private final Frame frame;
```
Lets update the documentation of this class mentioning it can be backed by frames.

cryptoe commented on processing/src/main/java/org/apache/druid/query/InlineDataSource.java:
```diff
@@ -53,14 +63,17 @@
   )
   {
-    this.rows = Preconditions.checkNotNull(rows, "'rows' must be nonnull");
+    this.rows = rows;
```
The inline DS can either be backed by a frame or by a rows. 
So lets error out if both are null.

paul-rogers replied:
If the code can use either rows or frames, but not both, then a cleaner design is to have two distinct classes. That avoids a bunch if if/then statements and the need for a reader to think through both cases at the same time.

cryptoe commented on processing/src/main/java/org/apache/druid/query/InlineDataSource.java:
```diff
@@ -190,15 +211,57 @@ public List<ColumnType> getColumnTypes()
+      return frameRows;
+    }
+
     return rows;
```
Nit: Lets add this in the else part ?

cryptoe commented on processing/src/main/java/org/apache/druid/query/InlineDataSource.java:
```diff
@@ -300,7 +363,7 @@ public boolean equals(Object o)
       return false;
     }
     InlineDataSource that = (InlineDataSource) o;
-    return rowsEqual(rows, that.rows) &&
```
maybe compare frames directly if possible and save on the cost of initializing various frameReaders and then de serializing them to rows.

LakshSingla replied:
I'll check if this works since that would be faster.
InlineDatasource's equality is done only in the tests afaik, therefore we should be fine if it doesn't work as such.

cryptoe commented on server/src/main/java/org/apache/druid/server/ClientQuerySegmentWalker.java:
```diff
@@ -565,10 +616,14 @@ private static <T, QueryType extends Query<T>> InlineDataSource toInlineDataSour
       final Sequence<T> results,
       final QueryToolChest<T, QueryType> toolChest,
       final AtomicInteger limitAccumulator,
```
we should never be using both limits I guess. Can we remove the additional 3 params and just pass another param called type ? and re-use the same limit variables ?

LakshSingla replied:
I am not sure of the use case. Theoretically, we can pass both the limits and error the query out if any one of them is reached. Is that the behavior we can encourage or do we want the user to give only one of the limit?

paul-rogers replied:
If we have two limits, we might want to give them names so we know which limit that `limitAccumulator` is accumulating: row count or memory bytes?

cryptoe commented on server/src/main/java/org/apache/druid/server/ClientQuerySegmentWalker.java:
```diff
@@ -577,25 +632,71 @@ private static <T, QueryType extends Query<T>> InlineDataSource toInlineDataSour
+    Frame frame = null;
+
+    // Try to serialize the results into a frame only if the memory limit is set on the server or the query
+    if (memoryLimitSet) {
```
We should only serialize the to frames only when memoryLimit Is set else the old code path should be invoked.

cryptoe commented on server/src/test/java/org/apache/druid/server/ClientQuerySegmentWalkerTest.java:
```diff
@@ -798,7 +798,36 @@ public void testTimeseriesOnGroupByOnTableErrorTooManyRows()
+
+
+  @Test
+  public void testTimeseriesOnGroupByOnTableErrorTooLarge()
```
I think tests around
1. String cols
2. Long cols
3. Double/float cols
4. Complex cols
5. Array cols
6. Nested cols 
Would help us build confidence in the feature. 

Some tests to checkout: 
* `CalciteQueryTests#testMaxSubqueryRows`
* `GroupByQueryRunnerTest#testGroupByMaxRowsLimitContextOverride`

Ideally all tests which have a subquery should be executed using the new code path but since its feature flagged it might not be a hard requirement.

paul-rogers commented on processing/src/main/java/org/apache/druid/query/InlineDataSource.java:
```diff
@@ -190,15 +211,57 @@ public List<ColumnType> getColumnTypes()
   @JsonIgnore
   public Iterable<Object[]> getRows()
   {
+    if (isBackedByFrame()) {
```
This is the kind of unnecessary complexity that comes when we have one class do two things, so would be cool to split this into three: base class, array implementation, frame implementation.

paul-rogers commented on processing/src/main/java/org/apache/druid/query/InlineDataSource.java:
```diff
@@ -300,7 +363,7 @@ public boolean equals(Object o)
     }
     InlineDataSource that = (InlineDataSource) o;
-    return rowsEqual(rows, that.rows) &&
+    return rowsEqual(getRowsAsList(), that.getRowsAsList()) &&
```
Does it even make semantic sense to compare two input sources for equality? Are we adding this become some static check told us we need it, but not because we actually use it?

paul-rogers commented on server/src/main/java/org/apache/druid/server/ClientQuerySegmentWalker.java:
```diff
@@ -577,25 +632,71 @@ private static <T, QueryType extends Query<T>> InlineDataSource toInlineDataSour
     final RowSignature signature = toolChest.resultArraySignature(query);
 
-    final ArrayList<Object[]> resultList = new ArrayList<>();
+    final List<Object[]> resultList = toolChest.resultsAsArrays(query, results).toList();
```
If we are under memory pressure, is it wise to materialize the entire result set? Can the results be processed incrementally in a streaming manner?

paul-rogers commented on server/src/main/java/org/apache/druid/server/ClientQuerySegmentWalker.java:
```diff
@@ -577,25 +632,71 @@ private static <T, QueryType extends Query<T>> InlineDataSource toInlineDataSour
+            new ArrayList<>()
+        );
+
+        final Cursor cursor = new InlineResultsCursor(resultList, signature);
```
All of this seems too complex to put inside the segment walker. For one thing, it is hard to test if it is an implementation detail. Perhaps pull out this logic into a separate class that can be unit tested extensively. For example, we'd want tests for hitting each of the limits, for handling variable-width columns, etc.

paul-rogers commented on server/src/main/java/org/apache/druid/server/ClientQuerySegmentWalker.java:
```diff
@@ -577,25 +632,71 @@
+        throw rlee;
+      }
+      catch (Exception e) {
+        log.info("Unable to write the subquery results to a frame. Results won't be accounted for in the memory"
+                 + "calculation");
```
Our error rules state that you must interpolate something, such as the reason why the subquery can't be written to a frame. And, yes, that code scanning item above is real.

cryptoe commented on processing/src/main/java/org/apache/druid/query/scan/ScanQueryQueryToolChest.java:
```diff
@@ -187,11 +199,75 @@ public RowSignature resultArraySignature(final ScanQuery query)
+            frame = Frame.wrap(frameWriter.toByteArray());
+          }
+
+          return new FrameSignaturePair(frame, result.getRowSignature());
```
We are making one frame per result sequence and each result sequence represents one segment. This does not seem very scalable. 
Lets leave a note here so that we can get back to this in a future PR.

cryptoe replied:
Since we can have different row sigs per segment, what we can do is only start a new frame when the row signature is different. This will reduce the number of frames by a lot.

cryptoe commented on processing/src/main/java/org/apache/druid/query/timeseries/TimeseriesQueryQueryToolChest.java:
```diff
@@ -451,6 +462,49 @@ public Sequence<Object[]> resultsAsArrays(
+      Sequence<Result<TimeseriesResultValue>> resultSequence
+  )
+  {
+    final RowSignature rowSignature = resultArraySignature(query);
```
Similar memory logic trick here as discussed in the scan query tool chest

cryptoe commented on processing/src/main/java/org/apache/druid/query/topn/TopNQueryQueryToolChest.java:
```diff
@@ -548,6 +561,46 @@ public Sequence<Object[]> resultsAsArrays(TopNQuery query, Sequence<Result<TopNR
+  @Override
+  public Sequence<FrameSignaturePair> resultsAsFrames(TopNQuery query, Sequence<Result<TopNResultValue>> resultSequence)
+  {
+    final RowSignature rowSignature = resultArraySignature(query);
```
Also mention the number of frames returned in each query type.

cryptoe commented on server/src/main/java/org/apache/druid/server/ClientQuerySegmentWalker.java:
```diff
@@ -163,12 +169,15 @@ public <T> QueryRunner<T> getQueryRunnerForIntervals(Query<T> query, Iterable<In
     final DataSource freeTradeDataSource = globalizeIfPossible(newQuery.getDataSource());
     // do an inlining dry run to see if any inlining is necessary, without actually running the queries.
     final int maxSubqueryRows = query.context().getMaxSubqueryRows(serverConfig.getMaxSubqueryRows());
+    final long maxSubqueryMemory = query.context().getMaxSubqueryMemoryBytes(serverConfig.getMaxSubqueryMemory());
```
Lets undocument the server property since it will enable all q's running with the new query path. 
Also we should benchmark the new query path.

cryptoe commented on processing/src/main/java/org/apache/druid/query/QueryToolChest.java:
```diff
@@ -330,4 +330,29 @@ public Sequence<Object[]> resultsAsArrays(QueryType query, Sequence<ResultType>
+   * this is done to preserve the space by not populating the column in that particular Frame and omitting it from its
+   * signature
+   */
+  public Sequence<FrameSignaturePair> resultsAsFrames(QueryType query, Sequence<ResultType> resultSequence)
```
Can we return an Optional<Sequence> and remove the `canFetchResultsAsFrames` call from the interface ?

LakshSingla replied:
Updated

cryptoe reviewed (commented):
Left some comments. 
Looking forward to the UT's.

LakshSingla commented on processing/src/main/java/org/apache/druid/query/FramesBackedInlineDataSource.java:
```diff
@@ -0,0 +1,239 @@
+    return rowSignature;
+  }
+
+  public List<Object[]> getRowsAsList()
```
Updated and refactored it further to reuse the `rowsAsSequence()`, so that it doesn't do any duplicate computation.

LakshSingla commented on processing/src/main/java/org/apache/druid/frame/write/FrameWriters.java:
```diff
@@ -58,19 +58,30 @@ public static FrameWriterFactory makeFrameWriterFactory(
     }
   }
 
+  public static FrameWriterFactory makeFrameWriterFactory(
```
Added this as a separate method originally because we don't require the boolean in the rest of the cases (i.e. the original ones in MSQ). Therefore it made sense to me to hide this complexity from the callers of this method that arent residing in the broker.

cryptoe commented on processing/src/main/java/org/apache/druid/query/scan/ConcatCursor.java:
```diff
@@ -0,0 +1,114 @@
+    if (currentCursor < cursors.size()) {
+      cursors.get(currentCursor).advance();
+      if (cursors.get(currentCursor).isDone()) {
+        ++currentCursor;
```
I think this is a common method. Maybe extract it out to `moveCursorIfDonetoNextNonEmpty()`

cryptoe commented on processing/src/test/java/org/apache/druid/query/scan/ConcatCursorTest.java:
```diff
@@ -0,0 +1,77 @@
+  @Test
+  public void testCursor()
+  {
+    Cursor dummyCursor1 = new ListCursor(new ArrayList<>());
```
Test case for
* empty cursors
* empty at end only 
* empty at start
* 2 empty cursors one after the other in all 3 places start , end , middle. 

Same set of test cases for advanceUninterruptibly. 

Maybe use parameterized tests.

cryptoe commented on processing/src/main/java/org/apache/druid/query/scan/ScanQueryQueryToolChest.java:
```diff
@@ -187,11 +203,112 @@ public RowSignature resultArraySignature(final ScanQuery query)
+    Yielder<ScanResultValue> yielder = Yielders.each(resultSequence);
+    RowSignature prevSignature = null;
+    List<Cursor> unwrittenCursors = null;
+    List<FrameSignaturePair> frameSignaturePairs = new ArrayList<>();
```
Can we add final here

cryptoe replied:
All of this is prematerialized. We need to make this lazy. For next() of the iterable, we always return one frame. That frame is generated lazily.

cryptoe commented on processing/src/main/java/org/apache/druid/query/scan/ScanQueryQueryToolChest.java:
```diff
@@ -187,11 +203,112 @@ public RowSignature resultArraySignature(final ScanQuery query)
 
+  /**
+   * This batches the fetched {@link ScanResultValue}s which have similar signatures and are consecutives. In best case
+   * it would return a single frame, and in the worst case, it would return as many frames as the number of results
```
```suggestion
   * it would return a single frame, and in the worst case, it would return as many frames as the number of {#link ScanResultValue}
```

cryptoe reviewed (commented):
Left some comments.

LakshSingla commented on processing/src/test/java/org/apache/druid/query/scan/ConcatCursorTest.java:
```diff
@@ -0,0 +1,77 @@
+    List<Object> tempList = new ArrayList<>();
+    // Initial iteration
+    while (!concatCursor.isDone()) {
+      tempList.add(concatCursor.getColumnSelectorFactory().makeColumnValueSelector("ignored").getObject());
```
I think this would be done implicitly when we assert the contents, since on each iteration we are adding the element to the list.

cryptoe commented on processing/src/main/java/org/apache/druid/query/scan/ScanQueryQueryToolChest.java:
```diff
@@ -217,77 +218,106 @@ public Optional<Sequence<FrameSignaturePair>> resultsAsFrames(
+      @Override
+      public boolean hasNext()
+      {
+        return !yielder.isDone() || nextScanResultValue != null;
```
This looks to me a impl of peeking iterator. `com.google.common.collect.PeekingIterator`

cryptoe commented on processing/src/main/java/org/apache/druid/query/scan/ScanQueryQueryToolChest.java:
```diff
@@ -217,77 +218,106 @@ public Optional<Sequence<FrameSignaturePair>> resultsAsFrames(
-      prevSignature = scanResultValue.getRowSignature();
-      yielder = yielder.next(null);
-    }
+        if (memoryLimitAccumulator != null) {
```
Nit: Please push memoryLimitAccumulator inside convertScanResultValuesToFrame

cryptoe commented on processing/src/main/java/org/apache/druid/query/scan/ScanQueryQueryToolChest.java:
```diff
@@ -187,11 +206,129 @@ public RowSignature resultArraySignature(final ScanQuery query)
+          final RowSignature nextRowSignature = scanResultValuePeekingIterator.peek().getRowSignature();
+          if (nextRowSignature != null && nextRowSignature.equals(rowSignature)) {
+            batch.add(scanResultValuePeekingIterator.next());
+          }
```
should there be an else here which breaks the while loop ?

cryptoe replied:
Please add tests cases in `ScanQueryQueryToolChestTest` for this method.

LakshSingla replied:
Good catch for the 1st comment, added test cases that encompass this as well other methods.

gianm reviewed (commented):
Other than the line comments, some high level stuff:

**Using `COMPLEX<json>`**

After seeing in https://github.com/apache/druid/pull/14139/files#r1183250939 that we need to suppress some `COMPLEX<json>` vs. traditional column equivalency tests due to them not currently passing, I don't think we should be using `COMPLEX<json>` automatically for unknown types right now.

Suggestion: keep the logic, but put it behind a special feature flag, so we can test it without it being always-on for the bytes-based-limit feature. Something like `useNestedForUnknownTypeInSubquery`. Undocumented, please, and default to `false`: its purpose is to help devs test things. If false, any unknown types mean we fall back to row-based limit rather than memory-based limit.

This change should be done in this PR, so the feature is safe as initially merged.

**Configuration**

Some additional thought is needed due to the change to not always use `COMPLEX<json>` for unknown types. It means that there will be cases where `maxSubqueryBytes` is configured, but we can't use it because we don't know the types and `useNestedForUnknownTypeInSubquery` is `false`. In this case we should fall back to `maxSubqueryRows`. I'm thinking this is a good way to handle it:

- If `maxSubqueryBytes` is not set, we use `maxSubqueryRows` for the limit.
- If `maxSubqueryBytes` is set, but we can't use bytes-based limiting (due to unknown types or some other reason), then we still use `maxSubqueryRows` for the limit.
- If `maxSubqueryBytes` is set, and we _can_ use bytes-based limiting, then we use `maxSubqueryBytes` for the limit and ignore `maxSubqueryRows`.

We should also have the configuration be as automatic as possible. Best thing that comes to mind is to dedicate some portion of the heap to subquery results, then apportion that amongst the maximum number of concurrent queries.

So, reasonable automatic setting for `maxSubqueryBytes` could be 50% of max heap, minus aggregate lookup size, divided by server thread pool size. (See `WorkerMemoryParameters#computeTotalLookupFootprint` for some code that computes lookup footprint.) For example, if heap is 4GB, with 100MB of lookups, and server thread pool size of 40, then an automatic subquery memory limit would be `(4GB - 100MB) / 40 = 97.5MB`.

Suggestion here:

- Default for `maxSubqueryBytes` is `"unlimited"` at first, meaning, well, unlimited.
- It can also be set to the string `"auto"` meaning we use the automatic calculation above. When the feature proves valuable & stable we would change the default to `"auto"`.
- It could also be set to a specific number.

Thoughts?

The configuration changes could be done in a follow-up patch that also adds documentation.

LakshSingla commented on processing/src/main/java/org/apache/druid/frame/write/FrameWriters.java:
```diff
@@ -53,24 +54,38 @@ private FrameWriters()
       final MemoryAllocatorFactory allocatorFactory,
-      final RowSignature signature,
-      final List<KeyColumn> sortColumns
+      @Nonnull final RowSignature signature,
```
Thanks for the pointer! Will fix the clutter.

LakshSingla commented on processing/src/main/java/org/apache/druid/query/QueryToolChest.java:
```diff
@@ -330,4 +331,32 @@ public Sequence<Object[]> resultsAsArrays(QueryType query, Sequence<ResultType>
+   * @param memoryLimitBytes Limit the memory results. Throws {@link ResourceLimitExceededException} if the result exceed
+   *                    the memoryLimitBytes
+   */
+  public Optional<Sequence<FrameSignaturePair>> resultsAsFrames(
```
I have updated the code with the suggestion here. 
While this is a much better way to tackle this and give a lot more autonomy to the callers, this suffers from one issue that we see with MSQ ingestions containing large data sketches, that Frame size is insufficient to hold a single row. We can set the limit higher here to something like 128MB, however, then it would mean that we would potentially be materializing 128MBs more than what we require (instead of fine graning it at a singular row level). 
Any ways to counteract this?

LakshSingla commented:
Thanks for the reviews!
Post the latest batch of changes, here are the major things that I have updated:
1. Create a `FrameBasedIndexedTable` to create an indexed table based on Frame datasource. This also changed the Frame type to COLUMNAR since to index the table we need to get COLUMNs corresponding to the indexed keys. 
2. Use a memory allocator factory to allow flexibility in converting the inline results to the frames
3. Added configurations, need to provide the fallback code paths as well.
4. Refactored the code and stylistic changes

cryptoe reviewed (approved):
Changes LGTM. 
Thanks @LakshSingla !!

abhishekagarwal87 commented on server/src/main/java/org/apache/druid/server/ClientQuerySegmentWalker.java:
```diff
@@ -613,96 +638,149 @@ private static <T, QueryType extends Query<T>> DataSource toInlineDataSource(
+      );
+    }
+    catch (Exception e) {
+      return Optional.empty();
```
why is this a debug log though? It should be WARN.

LakshSingla replied:
This will be executed per query so DEBUG is more appropriate according to me, otherwise, the logs will be cluttered with the exception message. Either we should:
1. Keep it as DEBUG info so that we don't have to see cluttered logs. This has the disadvantage that we won't be able to readily observe if we fallback to the default method/code
2. Don't catch the exception and let it propagate. The user will then report the issue and we can fix it.

2nd option means that there won't be a fallback in case we aren't able to convert it to frames. Since this is a newer feature, I think we should still have a fallback till we are confident that we can convert each query, and once it is more mature and the frames can handle array types (currently it can handle string arrays only), we can remove this fallback altogether and let the exception pass through.

abhishekagarwal87 replied:
Yeah. We shouldn't be doing 2nd option. We can do something like below
 - log the thing as Info. 
 - log the exception stack trace if debug is set in the query context.

abhishekagarwal87 commented on server/src/main/java/org/apache/druid/server/ClientQuerySegmentWalker.java:
```diff
@@ -613,96 +638,149 @@ private static <T, QueryType extends Query<T>> DataSource toInlineDataSource(
+          limitAccumulator.addAndGet(frame.getFrame().numRows());
+          if (memoryLimitAccumulator.addAndGet(frame.getFrame().numBytes()) >= memoryLimit) {
+            throw ResourceLimitExceededException.withMessage(
+                "Subquery generated results beyond maximum[%d] bytes",
```
whats the expected action from user?

LakshSingla replied:
Updating with a more appropriate error message

LakshSingla commented on processing/src/main/java/org/apache/druid/frame/segment/FrameCursorUtils.java:
```diff
@@ -67,4 +76,58 @@ public static Filter buildFilter(@Nullable Filter filter, Interval interval)
+              }
+
+              if (!firstRowWritten) {
+                throw new ISE("Row size is greater than the frame size.");
```
Yes this can be seen at the top level. Refactored with the DruidException and a more actionable error message

cryptoe commented:
Since this PR is liable to break due to merge conflicts, going ahead and merging this. 
@LakshSingla Please address the logging feedback from @abhishekagarwal87 as part of a separate PR.

cryptoe commented:
Thanks for the contribution @LakshSingla !!
